### PR TITLE
ENH: Add support for lstsq on stacks of matrices

### LINKED ISSUE #8720: ENH: broadcast lstsq
Many `linalg` functions are already broadcastable, `lstsq` isn't. A workaround is via `svd` which is already broadcasted:
```python
u, s, v = numpy.linalg.svd(A, full_matrices=False)
uTb = numpy.einsum('ijk,ij->ik', u, b)
xx = numpy.einsum('ijk, ij->ik', v, uTb / s)
```

### PR DESCRIPTION
**Pull request by eric-wieser**
Close #8720, at the cost of behavior changes in the `resids` return value. Marking as draft since I am publishing it primarily to facilitate discussion at that issue.

### CODE REVIEW

File: numpy/linalg/linalg.py [hunk 1]
```diff
@@ -2180,11 +2180,14 @@ def lstsq(a, b, rcond="warn"):
     If `a` is square and of full rank, then `x` (but for round-off error)
     is the "exact" solution of the equation.
 
+    .. versionchanged:: 1.19
+       Can now operate on stacks of matrices
+
     Parameters
     ----------
-    a : (M, N) array_like
+    a : (..., M, N) array_like
```
Comment (charris) [inline numpy/linalg/linalg.py]: Hmm, the current docstring here is very old and could use improvement. Just array_like on one line, then a description of the shapes in the explanatory part and how they are treated.

File: numpy/linalg/linalg.py [hunk 2]
```diff
@@ -2203,17 +2206,26 @@ def lstsq(a, b, rcond="warn"):
 
     Returns
     -------
-    x : {(N,), (N, K)} ndarray
+    x : {(N,), (..., N, K)} ndarray
         Least-squares solution. If `b` is two-dimensional,
         the solutions are in the `K` columns of `x`.
-    residuals : {(1,), (K,), (0,)} ndarray
+    residuals : {(), (..., K,)} ndarray
         Sums of residuals; squared Euclidean 2-norm for each column in
```
Comment (charris) [inline numpy/linalg/linalg.py]: Never much liked this, it isn't that useful. There have been a number of requests for returning the covariance matrix of the fitted coefficients.

**Comment by charris** [comment]
I'm going to suggest a new function here. I don't feel strongly about broadcasting either way, but there have been requests for the covariance matrix of the fitted parameters and I often end up computing that myself as it is more informative and allows for generating error bars on curve fits and such. Unfortunately, I can't see if DGELSD provides sufficient information to do that, the documentation simply says that the input design matrix is destroyed.  DGELSS looks like an option but I can't find it in our fallback routines. I suppose we could implement it for ourselves. Multiple return options would be required as computing the covariance could take significant time.

File: numpy/linalg/linalg.py (hunk 1 above)
Comment (eric-wieser) [reply numpy/linalg/linalg.py]: This style of docstring with the shape in the argument name is fairly prevalent throughout this module.

**Comment by eric-wieser** [comment]
That sounds like an orthogonal suggestion here - the only point of this change is to enable broadcasting, since `lstsq` is one of the few remaining unbroadcast functions in `np.linalg`.

Unfortunately, the meaning of `resids` has to be tweaked every so slightly to make those semantics possible.

File: numpy/linalg/linalg.py (hunk 1 above)
Comment (charris) [reply numpy/linalg/linalg.py]: Yep, could use improvement throughout. Not saying you should undertake the task in this PR.

**Comment by seberg** [comment]
It seemed to me the main change is that resids previously had (in some sense) a `keepdims` in there, while now it does not? I.e. you do not add a dimension of size 1 at the end anymore.
For all practical purposes that probably does not matter much hopefully and I guess without it maybe it generalizes slight less well.
Although, I am not immediately sure, is that change for the simplest call actually necessary (I bet you guys discussed it before)?

**Comment by eric-wieser** [comment]
> It seemed to me the main change is that resids previously had (in some sense) a keepdims in there, while now it does not? I.e. you do not add a dimension of size 1 at the end anymore.

It's a little worse than that. The old behavior was to return any of:

* A result of shape `(0,)` for an underconstrained solution (for any input shape)
* A result of shape `(1,)` for a well-constrained solution with input shape (M,)
* A result of shape `(K,)` for a well-constrained solution with input shape (M, K)

The new behavior is

* A result of shape `(..., N,)` for input shape (..., N, M)
* A result of shape `()` for input shape (M)

So there are two incompatible changes here, not just one.

**Comment by seberg** [comment]
I suppose I was guessing that the majority of use cases is likely currently either in the `(K,)` category (unchanged) or in the `(1,)` category which changed to `()`. The old underconstrained one seems strange, but of course it is not impossible someone actually checks `len(constr)` or so to see whether it was underconstrained... So in that sense, breaking the `(1,)` case may be better anyway, because at least it should break loudly when it breaks.

So the only option would be keeping the `(0,)` and `(1,)` shapes even though they are ugly... It is super hard to say how disruptive that change could be. It seems a question of boldness, this is the right thing if we are bold enough to risk breaking someone.
With release notes, I think I may be happy if we expect that almost all code that actually could break by this, should break loudly.

**Comment by charris** [comment]
> That sounds like an orthogonal suggestion here

Yes, I thought it might be more useful in the long run :) Computing the sum of squared residuals also has a noticeable impact on performance, which has also been the subject of complaints.

But back to the case at hand with some explanation for other reviewers.

The problem with the broadcast case it that the dimensions of the returned residual array must be determined by M, N, and K and cannot depend on runtime results like matrix rank. DEGLS will always compute residuals if `M > N`, regardless of if the matrix is full rank or not, there just might be some missing residuals if it isn't full rank, so I would just return the results from those residuals in that case, although that becomes problematical when there are few residuals. When `M = N` and the matrix is full rank, zeros seems the appropriate return. In all other cases NaN seems appropriate.

Unfortunately, we fixed the function rather than the documentation in #9986, so there is some question as to whether we should retain the NaN array return when the matrix rank is < N regardless.

The upshot is that we need to make the change to deal with stacked arrays. It is enough of a corner case that I think we can do that, having zero/NaN arrays should deal with the case of downstream users being caught without being aware that something has changed.

**Comment by charris** [comment]
In short pseudo code
```
if M > N:
    # There are residuals returned, use them
    # Note that this result should be divided by (M - N)
    # to get unbiased estimate of variance if the model is any good.
    return sum_of_squares
elseif rank(a) >= M:
   # No residuals returned, but they are zero
    return zeros
else:
   #  No residuals returned, but they exist
   return NaNs
```
And an array always returned for the residuals.

File: numpy/linalg/linalg.py [hunk 3]
```diff
@@ -2203,17 +2206,26 @@ def lstsq(a, b, rcond="warn"):
...
+    x : {(N,), (..., N, K)} ndarray
         Least-squares solution. If `b` is two-dimensional,
         the solutions are in the `K` columns of `x`.
-    residuals : {(1,), (K,), (0,)} ndarray
+    residuals : {(), (..., K,)} ndarray
         Sums of squared residuals: Squared Euclidean 2-norm for each column in
         ``b - a @ x``.
-        If the rank of `a` is < N or M <= N, this is an empty array.
-        If `b` is 1-dimensional, this is a (1,) shape array.
+        If the rank of `a` is < N or M <= N, this is filled with NaN.
```
Comment (eric-wieser) [inline numpy/linalg/linalg.py:2215]:
```suggestion
        If the rank of `a` is < N (which is always true if M < N), this is filled with NaN.
```

File: numpy/linalg/tests/test_linalg.py [hunk 4]
```diff
@@ -902,31 +902,31 @@ def test_0_size(self):
...
-            expect_resids = np.asarray(expect_resids)
-            if np.asarray(b).ndim == 1:
-                expect_resids.shape = (1,)
-                assert_equal(residuals.shape, expect_resids.shape)
-        else:
-            expect_resids = np.array([]).view(type(x))
+        expect_resids = np.where(
+            rank == n,
+            (abs(a @ x - b) ** 2).sum(axis=0),
+            np.nan  # lapack does not compute this for us, so we do not return it
```
Comment (eric-wieser) [inline numpy/linalg/tests/test_linalg.py:926]:
```suggestion
            # lapack does not compute this for us, so we do not return it
            np.nan
```

**Comment by rgommers** [comment]
@IvanYashchuk since you've just spent a bunch of time dealing with this `residuals` shape issue for PyTorch, would you mind having a look at this PR?

File: numpy/linalg/linalg.py [hunk 5]
```diff
@@ -2180,11 +2180,14 @@ def lstsq(a, b, rcond="warn"):
     Euclidean 2-norm :math:`||b - ax||`. If there are multiple minimizing 
     solutions, the one with the smallest 2-norm :math:`||x||` is returned.
 
+    .. versionchanged:: 1.19
```
Comment (IvanYashchuk) [inline numpy/linalg/linalg.py:2183]: Should this be `1.21.0`?

File: numpy/linalg/linalg.py [hunk 6]
```diff
@@ -2180,11 +2180,14 @@ def lstsq(a, b, rcond="warn"):
...
+       Can now operate on stacks of matrices
+
     Parameters
     ----------
-    a : (M, N) array_like
+    a : (..., M, N) array_like
         "Coefficient" matrix.
-    b : {(M,), (M, K)} array_like
+    b : {(M,), (..., M, K)} array_like
```
Comment (IvanYashchuk) [inline numpy/linalg/linalg.py:2190]: `numpy.linalg.solve` has the following description for `b` : `{(…, M,), (…, M, K)}, array_like`. `solve` works for the `(…, M,)` case, while `lstsq` from this PR raises `LinAlgError: Incompatible dimensions`.

Should the behavior of this function follow the `solve`?

**Review by IvanYashchuk — COMMENTED** [review]
The residuals change looks good to me. It's great to see the lstsq work on the stack of matrices.

I only have concerns about the behavior when the `b` array is a stack of vectors, `numpy.linalg.solve` works for this case, while `numpy.linalg.lstsq` doesn't, which is weird since they would compute the same result for the well-determined square case.

File: numpy/linalg/linalg.py (hunk 6 above)
Comment (eric-wieser) [reply numpy/linalg/linalg.py:2190]: That's probably a good idea, I hadn't realized that we can make that change without affecting compatibility. We'd replace the current `is_1d` logic with something like

https://github.com/numpy/numpy/blob/29561ed6bc96bc7472e43982e13bb96ae3c1a1f5/numpy/linalg/linalg.py#L384-L389

**Pull request closed without merging**